plotutils/graph.h

```diff
@@ -0,0 +1,205 @@
+// By convention, linemode #0 means no line at all (data points are disconnected).
+// 1: red, solid
+// 2: green, solid
+// 3: blue, solid
+// 4: magenta, solid
+// 5: cyan, solid
+// Linemodes #6 through #10 use the same five colors, but are dotted; linemodes #11 through #15 are dotdashed;
+// linemodes #16 through #20 are shortdashed; and linemodes #21 through #25 are longdashed.
+// So besides linemode #0, there are a total of 25 distinct colored linemodes
+enum class LineMode : int {
+  None = 0,
+};
+enum class LineColor : int { Red = 1, Green = 2, Blue = 3, Magenta = 4, Cyan = 5 };
+enum class LineStyle : int { Solid = 0, Dotted = 5, DotDashed = 10, ShortDashed = 15, LongDashed = 20 };
+inline LineMode CustomLineMode(LineColor color, LineStyle style) {
+  return static_cast<LineMode>(static_cast<int>(color) + static_cast<int>(style));
+}
+
+// https://www.gnu.org/software/plotutils/manual/en/html_node/Dataset-Options.html#Dataset-Options
+// The following table lists the first few symbols (by convention, symbol #0 means no symbol at all).
+// 1: dot
+// 2: plus (+)
+// 3: asterisk (*)
+// 4: circle
+// 5: cross
+enum class Symbol : int { No = 0, Dot = 1, Plus = 2, Asterisk = 3, Circle = 4, Cross = 5 };
+
+class Graph {
+ public:
+  Graph(const std::vector<std::pair<double, double>>& data) : data_(new SinglePlot(data)) {}
+  Graph(const std::vector<std::vector<std::pair<double, double>>>& data) : data_(new MultiPlot(data)) {}
+
+  Graph& X(const std::string& x_label) {
+    parameters_["-X"] = SomewhatEscape(x_label);
+    return *this;
+  }
+
+  Graph& Y(const std::string& y_label) {
+    parameters_["-Y"] = SomewhatEscape(y_label);
+    return *this;
+  }
+
+  Graph& Label(const std::string& label) {
+    parameters_["-L"] = SomewhatEscape(label);
+    return *this;
+  }
+
+  Graph& LineWidth(double w) {
+    parameters_["-W"] = bricks::strings::to_string(w);
+    return *this;
+  }
+
+  Graph& BitmapSize(int x, int y = 0) {
+    assert(x > 0);
+    assert(y >= 0);
+    parameters_["--bitmap-size"] = bricks::strings::Printf("%dx%d", x, y ? y : x);
+    return *this;
+  }
+
+  Graph& GridStyle(GridStyle grid_style) {
+    parameters_["-g"] = bricks::strings::to_string(static_cast<int>(grid_style));
+    return *this;
+  }
+
+  Graph& LineMode(LineMode line_mode) {
+    parameters_["-m"] = bricks::strings::to_string(static_cast<int>(line_mode));
+    return *this;
+  }
+
+  Graph& Symbol(Symbol symbol, double size = 0) {
+    parameters_["-S"] = bricks::strings::to_string(static_cast<int>(symbol));
+    if (size) {
+      parameters_["-S"] += ' ' + bricks::strings::to_string(size);
+    }
+    return *this;
+  }
+
+  Graph& OutputFormat(const std::string& output_format) {
+    parameters_["-T"] = output_format;
+    return *this;
+  }
+
+  operator std::string() const {
+    const std::string input_file_name = bricks::FileSystem::GenTmpFileName();
```

Comment:
If you meant ```popen()```, it's one-directional and Linux-specific. I thought, if at least one temp file is there to create, I might as well create two.